File: go/word_reverser/word_reverser_3.go
// word_reverser.go
// This program reads a string input from the user, reverses the words in the string, and prints the reversed string.
// The purpose of this program is to help programmers understand how to manipulate strings and handle user input in Go.
// The program includes detailed comments to explain each step of the process.

package main

import (
    "bufio"
    "fmt"
    "os"
    "strings"
)

// reverseWords takes a string and returns a new string with the words in reverse order.
func reverseWords(input string) string {
    // Split the input string into a slice of words.
    words := strings.Fields(input)
    // Initialize an empty slice to hold the reversed words.
    reversedWords := make([]string, 0, len(words))
    // Iterate over the words slice in reverse order and append each word to the reversedWords slice.
    for i := len(words) - 1; i >= 0; i-- {
        reversedWords = append(reversedWords, words[i])
    }
    // Join the reversedWords slice into a single string with spaces between the words.
    return strings.Join(reversedWords, " ")
}

func main() {
    // Create a new reader to read input from the standard input (console).
    reader := bufio.NewReader(os.Stdin)
    // Prompt the user to enter a string.
    fmt.Print("Enter a string to reverse the words: ")
    // Read the input string from the user.
    input, _ := reader.ReadString('\n')
    // Trim any leading or trailing whitespace from the input string.
    input = strings.TrimSpace(input)
    // Reverse the words in the input string.
    reversed := reverseWords(input)
    // Print the reversed string.
    fmt.Println("Reversed string:", reversed)
    // Note: The program does not explicitly close the reader, which may lead to a resource leak.
}

File: kotlin/word_reverser/word_reverser_3.kt
/**
 * This program is designed to reverse the words in a given sentence.
 * It takes a sentence as input and outputs the sentence with the words reversed.
 * The purpose of this program is to provide a detailed example of how to manipulate strings in Kotlin.
 * It includes verbose comments to explain each step of the process.
 */

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    // Create a BufferedReader to read input from the user
    val reader = BufferedReader(InputStreamReader(System.`in`))

    // Prompt the user to enter a sentence
    println("Please enter a sentence:")

    // Read the input sentence from the user
    val sentence = reader.readLine()

    // Split the sentence into words
    val words = sentence.split(" ")

    // Create a variable to store the reversed sentence
    var reversedSentence = ""

    // Iterate over the words in reverse order
    for (i in words.indices.reversed()) {
        // Append each word to the reversed sentence
        reversedSentence += words[i] + " "
    }

    // Print the reversed sentence
    println("Reversed sentence: $reversedSentence")

    // Close the BufferedReader
    reader.close()
}

